pyuvdata/uvdata/mir.py

```diff
@@ -0,0 +1,194 @@
+    writing Mir files. This class should not be interacted with directly,
+    instead use the read_mir and write_mir methods on the UVData class.
+    """
+
+    def read_mir(self, filepath, isource=None, irec=None, isb=None, corrchunk=None):
+        """
+        Read in data from an SMA MIR file, and map to the UVData model.
+
+        Note that with the exception of filename, the reset of the parameters are
+        used to sub-select a range of data that matches the limitations of the current
+        instantiation of pyuvdata  -- namely 1 spectral window, 1 source. These could
+        be dropped in the future, as pyuvdata capabilities grow.
+
+        Parameters
+        ----------
+        filepath : str
+             The file path to the MIR folder to read from.
+        isource : int
+            Source code for MIR dataset
+        irec : int
+            Receiver code for MIR dataset
+        isb : int
+            Sideband code for MIR dataset
+        corrchunk : int
+            Correlator chunk code for MIR dataset
+        """
+        sma_lat = 0.3459976585365961
+        sma_lon = -2.713594675620429
+        sma_alt = 4080.0
+
+        # Use the mir_parser to read in metadata, which can be used to select data.
+        mir_data = mir_parser.MirParser(filepath)
+
+        # Select out data that we want to work with.
+        if isource is None:
+            isource = mir_data.in_read["isource"][0]
+        if irec is None:
+            irec = mir_data.bl_read["irec"][0]
+        if isb is None:
+            isb = mir_data.bl_read["isb"][0]
+        if corrchunk is None:
+            corrchunk = mir_data.sp_read["corrchunk"][0]
+
+        mir_data.use_in = mir_data.in_read["isource"] == isource
+        mir_data.use_bl = np.logical_and(
+            np.logical_and(
+                mir_data.bl_read["isb"] == isb, mir_data.bl_read["ipol"] == 0
+            ),
+            mir_data.bl_read["irec"] == irec,
+        )
+        mir_data.use_sp = mir_data.sp_read["corrchunk"] == corrchunk
+
+        # Load up the visibilities into the MirParser object. This will also update the
+        # filters, and will make sure we're looking at the right metadata.
+        mir_data._update_filter()
+        if len(mir_data.in_data) == 0:
+            raise IndexError("No valid records matching those selections!")
+
+        mir_data.load_data(load_vis=True, load_raw=True)
+
+        # Create a simple array/list for broadcasting values stored on a
+        # per-intergration basis in MIR into the (tasty) per-blt records in UVDATA.
+        bl_in_maparr = [mir_data.inhid_dict[idx] for idx in mir_data.bl_data["inhid"]]
+
+        # Derive Nants_data this value from baselines.
+        self.Nants_data = len(
+            np.unique(
+                np.concatenate((mir_data.bl_data["iant1"], mir_data.bl_data["iant2"]))
+            )
+        )
+
+        self.Nants_telescope = 8
+        self.Nbls = int(self.Nants_data * (self.Nants_data - 1) / 2)
```

Comment:
So the array has 8 antennas in total -- in the past, we used to be able to fold in other telescopes (i.e., JCMT and CSO [rip]). But in the "modern era" (which we're defining as having started with our ROACH2-based correlator), we only have the ability to record data for just those 8 antennas, and that's unlikely to change anytime in the next decade. We decided early on when going down this route that we were only going to support data from this modern era, mostly because of a lot of one-off issues with the older data that we just don't have the manpower to fix.

We do actually record auto-correlations -- I just hadn't thought of integrating them into the data. They're not very well-integrated into our data structure, so folding them in cleanly will require a bit of thought. Since they are very rarely used (at present, I seem to be the only person on the project who even looks at them), I might wait to include this feature on a following release, if that's alright.

Reply:
Thanks very much for the explanation. That sounds fair to me.